components/local-app/pkg/localapp/run.go

```diff
@@ -0,0 +1,325 @@
+		logrus.WithError(closeErr).Error("server connection failed")
+		os.Exit(1)
+	})
+	if err != nil {
+		return err
+	}
+
+	cb := bastion.CompositeCallbacks{
+		&logCallbacks{},
+	}
+
+	s := &bastion.SSHConfigWritingCallback{Path: opts.sshConfigPath}
+	if opts.sshConfigPath != "" {
+		cb = append(cb, s)
+	}
+
+	b = bastion.New(client, opts.localAppTimeout, cb)
+	b.EnableAutoTunnel = opts.autoTunnel
+	grpcServer := grpc.NewServer()
+	appapi.RegisterLocalAppServer(grpcServer, bastion.NewLocalAppService(b, s))
+	allowOrigin := func(origin string) bool {
+		// Is the origin a subdomain of the installations hostname?
+		return hostRegex.Match([]byte(origin))
+	}
+	go func() {
+		err := http.ListenAndServe("localhost:"+strconv.Itoa(opts.apiPort), grpcweb.WrapServer(grpcServer,
+			grpcweb.WithCorsForRegisteredEndpointsOnly(false),
+			grpcweb.WithOriginFunc(allowOrigin),
+			grpcweb.WithWebsockets(true),
+			grpcweb.WithWebsocketOriginFunc(func(req *http.Request) bool {
+				origin, err := grpcweb.WebsocketRequestOrigin(req)
+				if err != nil {
+					return false
+				}
+				return allowOrigin(origin)
+			}),
+			grpcweb.WithWebsocketPingInterval(15*time.Second),
+		))
+		if err != nil {
+			logrus.WithError(err).Error("API endpoint failed to start")
+			os.Exit(1)
+		}
+	}()
+	defer grpcServer.Stop()
+	return b.Run()
+}
+
+func connectToServer(loginOpts auth.LoginOpts, reconnectionHandler func(), closeHandler func(error)) (*gitpod.APIoverJSONRPC, error) {
+	var client *gitpod.APIoverJSONRPC
+	onClose := func(closeErr error) {
+		if client != nil {
+			closeHandler(closeErr)
+		}
+	}
+	tkn, err := auth.GetToken(loginOpts.GitpodURL)
+	if err != nil {
+		return nil, err
+	}
+
+	if tkn != "" {
+		// try to connect with existing token
+		client, err = tryConnectToServer(loginOpts.GitpodURL, tkn, reconnectionHandler, onClose)
+		if client != nil {
+			return client, err
+		}
+		_, invalid := err.(*auth.ErrInvalidGitpodToken)
+		if !invalid {
+			return nil, err
+		}
+		// existing token is invalid, try again
+		logrus.WithError(err).WithField("origin", loginOpts.GitpodURL).Error()
+	}
+
+	tkn, err = login(loginOpts)
+	if err != nil {
+		return nil, err
+	}
+	client, err = tryConnectToServer(loginOpts.GitpodURL, tkn, reconnectionHandler, onClose)
+	return client, err
+}
+
+func tryConnectToServer(gitpodUrl string, tkn string, reconnectionHandler func(), closeHandler func(error)) (*gitpod.APIoverJSONRPC, error) {
+	wshost := gitpodUrl
+	wshost = strings.ReplaceAll(wshost, "https://", "wss://")
+	wshost = strings.ReplaceAll(wshost, "http://", "ws://")
+	wshost += "/api/v1"
+	client, err := gitpod.ConnectToServer(wshost, gitpod.ConnectToServerOpts{
+		Context:             context.Background(),
+		Token:               tkn,
+		Log:                 logrus.NewEntry(logrus.StandardLogger()),
+		ReconnectionHandler: reconnectionHandler,
+		CloseHandler:        closeHandler,
+		ExtraHeaders: map[string]string{
+			"User-Agent":       "gitpod/local-companion",
```

Comment:
should be different client

Reply:
Actually not, because this is only used for the local app and not the CLI.